server/src/main/java/org/elasticsearch/common/util/set/Sets.java

```diff
@@ -165,11 +175,12 @@ public static <T> Set<T> intersection(Set<T> set1, Set<T> set2) {
      *
      * @param set      set to copy
      * @param elements elements to add
+     * @return unmodifiable copy of the input set with the extra elements added
      */
     @SuppressWarnings("unchecked")
     public static <E> Set<E> addToCopy(Set<E> set, E... elements) {
         final var res = new HashSet<>(set);
         Collections.addAll(res, elements);
-        return Set.copyOf(res);
+        return Collections.unmodifiableSet(res);
```

Comment:
Makes sense. But since we are here, what about building res with the correct initial capacity, and then use HashSet#addAll (twice)? Should avoid potential re-allocations due to a larger `elements` size. Wdyt?

Reply:
It could happen that all / most 'elements' are in the 'set' already. And then we'd be overallocating the 'res' capacity.
If you still think pre-sized 'res' would be better I'll do that.

Reply:
No, that makes sense, it's a concrete possibility. Let's keep it the current way.

Reply:
IIRC Set.copyOf was specifically chosen here so as to avoid the extra indirection Collections.unmodifiableSet produces. You can see most of the uses are static. If we did call this multiple times, unmodifiableSet would create an extra level of indirection, each successive call making access require more indirections. I think this should stay Set.copyOf.

Reply:
Thank you for this remark.
You are right, the calls are static, so there will be a limited number of "entries copying". But the resulting set might be used a lot of times, "paying the indirection cost" on each call.
I will revert this change.